### ISSUE TITLE
optipng-bin failed to install on windows 7

### ISSUE THREAD

**Issue by aaronward** [imagemin/optipng-bin#58]
I'm attempting to install grunt-contrib-imagemin. I'm running all the commands from the command line tool running as Administrator. When installing, I get the following output:

```javascript

$ npm install grunt-contrib-imagemin --save-dev
npm WARN package.json sample-test@1.0.0 No description
npm WARN package.json sample-test@1.0.0 No repository field.
npm WARN package.json sample-test@1.0.0 No README data
|


> optipng-bin@3.0.2 postinstall C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\grunt-contrib-imagemin\node_modules\imagemin\node_module
s\imagemin-optipng\node_modules\optipng-bin
> node lib/install.js

  ‼ Command failed: C:\Windows\system32\cmd.exe /s /c "C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\grunt-contrib-imagemin\node_modul
es\imagemin\node_modules\imagemin-optipng\node_modules\optipng-bin\vendor\optipng.exe --version"
'C:\Users\aaron.ward\Downloads\VZCON-2806' is not recognized as an internal or external command,
operable program or batch file.

  ‼ optipng pre-build test failed
  i compiling from source
  × Error: ./configure --with-system-zlib --prefix="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\grunt-contrib-imagemin\node_modules\
imagemin\node_modules\imagemin-optipng\node_modules\optipng-bin\vendor" --bindir="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\grunt-
contrib-imagemin\node_modules\imagemin\node_modules\imagemin-optipng\node_modules\optipng-bin\vendor" && make install
Command failed: C:\Windows\system32\cmd.exe /s /c "./configure --with-system-zlib --prefix="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modu
les\grunt-contrib-imagemin\node_modules\imagemin\node_modules\imagemin-optipng\node_modules\optipng-bin\vendor" --bindir="C:\Users\aaron.ward\Downloads\VZCON-2806
HTML\VZCON-2806 HTML\node_modules\grunt-contrib-imagemin\node_modules\imagemin\node_modules\imagemin-optipng\node_modules\optipng-bin\vendor""
'.' is not recognized as an internal or external command,
operable program or batch file.

    at ChildProcess.exithandler (child_process.js:744:12)
    at ChildProcess.emit (events.js:110:17)
    at maybeClose (child_process.js:1008:16)
    at Socket.<anonymous> (child_process.js:1176:11)
    at Socket.emit (events.js:107:17)
    at Pipe.close (net.js:476:12)

> jpegtran-bin@3.0.4 postinstall C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\grunt-contrib-imagemin\node_modules\imagemin\node_modul
es\imagemin-jpegtran\node_modules\jpegtran-bin
> node lib/install.js

  √ jpegtran pre-build test passed successfully

> gifsicle@3.0.1 postinstall C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\grunt-contrib-imagemin\node_modules\imagemin\node_modules\i
magemin-gifsicle\node_modules\gifsicle
> node lib/install.js

  √ gifsicle pre-build test passed successfully
grunt-contrib-imagemin@0.9.4 node_modules\grunt-contrib-imagemin
├── gulp-rename@1.2.2
├── async@0.9.2
├── chalk@1.1.0 (escape-string-regexp@1.0.3, ansi-styles@2.1.0, supports-color@2.0.0, strip-ansi@3.0.0, has-ansi@2.0.0)
├── pretty-bytes@1.0.4 (get-stdin@4.0.1, meow@3.3.0)
└── imagemin@3.2.0 (get-stdin@4.0.1, path-exists@1.0.0, optional@0.1.3, buffer-to-vinyl@1.0.0, through2@0.6.5, meow@3.3.0, concat-stream@1.5.0, stream-combiner2@1.
0.2, vinyl-fs@1.0.0, imagemin-svgo@4.1.2, imagemin-optipng@4.3.0, imagemin-jpegtran@4.3.0, imagemin-gifsicle@4.2.0)
```

Thinking it might have been the package itself, I uninstalled grunt-contrib-imagemin and installed the optipng-bin package directly. Those errors are here:

```javascript

$ npm install --save optipng-bin
npm WARN package.json sample-test@1.0.0 No description
npm WARN package.json sample-test@1.0.0 No repository field.
npm WARN package.json sample-test@1.0.0 No README data
-
> optipng-bin@3.0.2 postinstall C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\optipng-bin
> node lib/install.js

  ‼ Command failed: C:\Windows\system32\cmd.exe /s /c "C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\optipng-bin\vendor\optipng.exe --version"
'C:\Users\aaron.ward\Downloads\VZCON-2806' is not recognized as an internal or external command,
operable program or batch file.

  ‼ optipng pre-build test failed
  i compiling from source
  × Error: ./configure --with-system-zlib --prefix="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\optipng-bin\vendor" --bindir="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\optipng-bin\vendor" && make install
Command failed: C:\Windows\system32\cmd.exe /s /c "./configure --with-system-zlib --prefix="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\optipng-bin\vendor" --bindir="C:\Users\aaron.ward\Downloads\VZCON-2806 HTML\VZCON-2806 HTML\node_modules\optipng-bin\vendor""
'.' is not recognized as an internal or external command,
operable program or batch file.

    at ChildProcess.exithandler (child_process.js:744:12)
    at ChildProcess.emit (events.js:110:17)
    at maybeClose (child_process.js:1008:16)
    at Process.ChildProcess._handle.onexit (child_process.js:1080:5)
optipng-bin@3.0.2 node_modules\optipng-bin
├── logalot@2.1.0 (figures@1.3.5, squeak@1.2.0)
├── bin-wrapper@3.0.2 (lazy-req@1.0.0, os-filter-obj@1.0.3, each-async@1.1.1, bin-check@2.0.0, bin-version-check@2.1.0, download@4.1.4)
└── bin-build@2.1.2 (exec-series@1.0.2, url-regex@2.1.3, tempfile@1.1.1, archive-type@2.1.0, rimraf@2.4.2, download@4.1.4, decompress@2.3.0)

```

I'm trying to set this up while doing a few tutorials on image compression. What might I be missing on installing this package? All other packages I have installed have been very straight-forward. While I have noticed some other having similar issues, I'm not behind any type of firewall or proxy...

**Comment by alexpaluzzi** [comment]
Yeah, this is a major headache. Been this way for the last month now. Can't do anything.

**Comment by testn** [comment]
Ran into the same issue on Mac as well

**Comment by danielupstone** [comment]
Looks like fundamentally this is an issue with node's child_proccess.spawn() and .exec() not handling spaces in paths / quotes well for windows.

From what I've read, this could be worked around with the cwd option for spawn.
https://github.com/kevva/bin-wrapper and https://github.com/kevva/bin-check would need to add support for options and not joining the dest and use values.

**Comment by SamuelMarks** [comment]
In the meantime, how about looking for `optipng` in `PATH` [`stat`] first, and if it is installed, use that system version?

That way we can still use packages which depend on `optipng`, by installing the binary manually.

Yes it'd be a hack, but at least it'd make things work!

**Comment by shift-evgeny** [comment]
This is still a problem every time optipng-bin is upgraded. If this is a Node bug like @danielupstone says, is there an issue in the Node bug tracker about it?

**Comment by shinnn** [comment]
Fixed in optipng-bin v3.0.4.
